decay.go

```diff
@@ -0,0 +1,268 @@
+
+	// bumpCh queues bump commands to be processed by the loop.
+	bumpCh chan bumpCmd
+
+	// closure thingies.
+	closeCh chan struct{}
+	doneCh  chan struct{}
+	err     error
+}
+
+var _ connmgr.Decayer = (*decayer)(nil)
+
+// DecayerCfg is the configuration object for the Decayer.
+type DecayerCfg struct {
+	Resolution time.Duration
+	Clock      clock.Clock
+}
+
+// WithDefaults writes the default values on this DecayerConfig instance,
+// and returns itself for chainability.
+//
+//  cfg := (&DecayerCfg{}).WithDefaults()
+//  cfg.Resolution = 30 * time.Second
+//  t := NewDecayer(cfg, cm)
+func (cfg *DecayerCfg) WithDefaults() *DecayerCfg {
+	cfg.Resolution = DefaultResolution
+	return cfg
+}
+
+// NewDecayer creates a new decaying tag registry.
+func NewDecayer(cfg *DecayerCfg, mgr *BasicConnMgr) (*decayer, error) {
+	// use real time if the Clock in the config is nil.
+	if cfg.Clock == nil {
+		cfg.Clock = clock.New()
+	}
+
+	d := &decayer{
+		cfg:       cfg,
+		mgr:       mgr,
+		clock:     cfg.Clock,
+		knownTags: make(map[string]*decayingTag),
+		bumpCh:    make(chan bumpCmd, 128),
+		closeCh:   make(chan struct{}),
+		doneCh:    make(chan struct{}),
+	}
+
+	// kick things off.
+	go d.process()
+
+	return d, nil
+}
+
+// decayingTag represents a decaying tag, with an associated decay interval, a
+// decay function, and a bump function.
+type decayingTag struct {
+	trkr      *decayer
+	name      string
+	interval  time.Duration
+	nextRound uint64
+	decayFn   connmgr.DecayFn
+	bumpFn    connmgr.BumpFn
+}
+
+func (d *decayer) RegisterDecayingTag(name string, interval time.Duration, decayFn connmgr.DecayFn, bumpFn connmgr.BumpFn) (connmgr.DecayingTag, error) {
+	d.tagsMu.Lock()
+	defer d.tagsMu.Unlock()
+
+	tag, ok := d.knownTags[name]
+	if ok {
+		return nil, fmt.Errorf("decaying tag with name %s already exists", name)
+	}
+
+	if interval < d.cfg.Resolution {
+		err := fmt.Errorf("decay interval for %s (%s) is lower than tracker's resolution (%s)",
+			name,
+			interval,
+			d.cfg.Resolution)
+		return nil, err
+	}
+
+	if interval%d.cfg.Resolution != 0 {
+		err := fmt.Errorf("decay interval for tag %s (%s) is not a multiple of tracker's resolution (%s); "+
+			"preventing undesired loss of precision",
+			name,
+			interval,
+			d.cfg.Resolution)
+		return nil, err
+	}
+
+	initRound := atomic.LoadUint64(&d.currRound) + uint64(interval/d.cfg.Resolution)
+	tag = &decayingTag{
+		trkr:      d,
+		name:      name,
+		interval:  interval,
+		nextRound: initRound,
+		decayFn:   decayFn,
+		bumpFn:    bumpFn,
+	}
+
+	d.knownTags[name] = tag
+	return tag, nil
+}
+
+// Close closes the Decayer. It is idempotent.
+func (d *decayer) Close() error {
+	select {
+	case <-d.doneCh:
+		return d.err
+	default:
+	}
+
+	close(d.closeCh)
+	<-d.doneCh
+	return d.err
+}
+
+// process is the heart of the tracker. It performs the following duties:
+//
+//  1. Manages decay.
+//  2. Applies score bumps.
+//  3. Yields when closed.
+func (d *decayer) process() {
```

Comment:
A performance consideration here is that this loop might run way too often if the resolution is small.

Reply:
True, maybe we should introduce a hard lower limit for the resolution?

Reply:
Yeah, that would work.